test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/BindToStreamTests.cs

```diff
@@ -0,0 +1,461 @@
+            {
+                this.Access = access;
+            }
+        }
+
+        // Test that leaving an out-parameter as null does not create a stream.
+        // This means it shouldn't even call the Attribute-->Stream converter function. 
+        [Fact]
+        public void NullOutParamDoesNotWriteStream()
+        {
+            TestWorker<ConfigNullOutParam>();
+        }
+
+        public class ConfigNullOutParam : IExtensionConfigProvider, ITest<ConfigNullOutParam>,
+            IConverter<TestStreamAttribute, Stream>
+        {
+            public void Initialize(ExtensionConfigContext context)
+            {
+                context.AddBindingRule<TestStreamAttribute>().
+                    BindToStream(this, FileAccess.ReadWrite);
+            }
+
+            public void Test(TestJobHost<ConfigNullOutParam> host)
+            {
+                host.Call("WriteString");
+                // Convert was never called 
+            }
+
+            public Stream Convert(TestStreamAttribute input)
+            {
+                // Should never even try to create a stream when 'out T' is set to null.
+                throw new InvalidOperationException("Test cases should never create stream");
+            }
+
+            // Explicit Write access 
+            public void WriteString(
+                [TestStream] out string x
+                )
+            {
+                x = null; // Don't write anything 
+            }
+        }
+
+        // Verify that BindToStream rule still honors [AutoResolve] on the attribute properties.
+        [Fact]
+        public void TestAutoResolve()
+        {
+            TestWorker<ConfigAutoResolve>();
+        }
+
+        public class ConfigAutoResolve : IExtensionConfigProvider, ITest<ConfigAutoResolve>,
+            IConverter<TestStreamAttribute, Stream>
+        {
+            private string _log;
+
+            public void Initialize(ExtensionConfigContext context)
+            {
+                context.AddBindingRule<TestStreamAttribute>().
+                    BindToStream(this, FileAccess.ReadWrite);
+            }
+
+            public void Test(TestJobHost<ConfigAutoResolve> host)
+            {
+                host.Call("Read", new { x = 456 });
+                // Convert was never called 
+
+                Assert.Equal("456-123", _log);
+            }
+
+            public Stream Convert(TestStreamAttribute input)
+            {
+                if (input.Access == FileAccess.Read)
+                {
+                    var value = input.Path; // Will exercise the [AutoResolve]
+                    var stream = new MemoryStream(Encoding.UTF8.GetBytes(value));
+                    stream.Position = 0;
+                    return stream;
+                }
+                throw new InvalidOperationException();
+            }
+
+            // Explicit Write access 
+            public void Read(
+                [TestStream(Path="{x}-%y%")] string value
```

Comment:
Where does the %y% come from?

Reply:
%...% means "take from appsettings". The test wired up an appsetting converter to substitute  this. 

---
In reply to: [145225541](https://github.com/Azure/azure-webjobs-sdk/pull/1387#discussion_r145225541) [](ancestors = 145225541)